pkg/controllers/sandbox/unsandboxsubcluster_reconciler.go

```diff
@@ -0,0 +1,267 @@
+/*
+ (c) Copyright [2021-2024] Open Text.
+ Licensed under the Apache License, Version 2.0 (the "License");
+ You may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package sandbox
+
+import (
+	"context"
+	"errors"
+
+	"github.com/go-logr/logr"
+	vutil "github.com/vertica/vcluster/vclusterops/util"
+	vapi "github.com/vertica/vertica-kubernetes/api/v1"
+	"github.com/vertica/vertica-kubernetes/pkg/controllers"
+	vdbcontroller "github.com/vertica/vertica-kubernetes/pkg/controllers/vdb"
+	"github.com/vertica/vertica-kubernetes/pkg/events"
+	vmeta "github.com/vertica/vertica-kubernetes/pkg/meta"
+	"github.com/vertica/vertica-kubernetes/pkg/vadmin"
+	"github.com/vertica/vertica-kubernetes/pkg/vadmin/opts/unsandboxsc"
+	"github.com/vertica/vertica-kubernetes/pkg/vdbstatus"
+	corev1 "k8s.io/api/core/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// UnsandboxSubclusterReconciler will unsandbox subclusters in the sandbox of a sandbox config map
+type UnsandboxSubclusterReconciler struct {
+	SRec *SandboxConfigMapReconciler
+	Vdb  *vapi.VerticaDB
+	Log  logr.Logger
+	client.Client
+	Dispatcher  vadmin.Dispatcher
+	PFacts      *vdbcontroller.PodFacts
+	ConfigMap   *corev1.ConfigMap
+	InitiatorIP string // The IP of the pod that we run vclusterOps from
+}
+
+func MakeUnsandboxSubclusterReconciler(r *SandboxConfigMapReconciler, vdb *vapi.VerticaDB, log logr.Logger,
+	cli client.Client, pfacts *vdbcontroller.PodFacts, dispatcher vadmin.Dispatcher, configMap *corev1.ConfigMap) controllers.ReconcileActor {
+	pfactsForMainCluster := pfacts.Copy(vapi.MainCluster)
+	return &UnsandboxSubclusterReconciler{
+		SRec:       r,
+		Log:        log.WithName("SandboxControllerUnsandboxSubclusterReconciler"),
+		Vdb:        vdb,
+		Client:     cli,
+		Dispatcher: dispatcher,
+		PFacts:     &pfactsForMainCluster,
+		ConfigMap:  configMap,
+	}
+}
+
+// Reconcile will unsandbox subclusters, remove the sandbox config map, and update sandbox status of vdb
+func (r *UnsandboxSubclusterReconciler) Reconcile(ctx context.Context, _ *ctrl.Request) (ctrl.Result, error) {
+	// no-op for ScheduleOnly init policy or enterprise db
+	if r.Vdb.Spec.InitPolicy == vapi.CommunalInitPolicyScheduleOnly || !r.Vdb.IsEON() {
+		return ctrl.Result{}, nil
+	}
+
+	// collect pod facts for main cluster
+	if err := r.PFacts.Collect(ctx, r.Vdb); err != nil {
```

Comment:
Can we move the collection to spot where we need to know its needed? For instance, if we come in here and there is no unsandboxing to do, we should not collect pod facts.

Reply:
We need the podfacts to correct the sandbox status. The function right below it needs the pod facts.